Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -133,34 +141,63 @@ internal void ValidateSecurePassword(SecureString password)
             password == null ? string.Empty : SecureStringHelper.Decode(password);
 
-        internal SFSession GetSession(string connStr, SecureString password)
+        internal SFSession GetSession(string connStr, SecureString password, SecureString passcode)
         {
             s_logger.Debug("SessionPool::GetSession" + PoolIdentification());
+            SFSession session = null;
+            var sessionProperties = SFSessionProperties.ParseConnectionString(connStr, password);
+            ValidatePoolingIfPasscodeProvided(sessionProperties);
             if (!GetPooling())
-                return NewNonPoolingSession(connStr, password);
+                return NewNonPoolingSession(connStr, password, passcode);
             var sessionOrCreateTokens = GetIdleSession(connStr);
+            if(sessionProperties.TryGetValue(SFSessionProperty.AUTHENTICATOR, out var authenticator) && authenticator == MFACacheAuthenticator.AUTH_NAME)
+                session = sessionOrCreateTokens.Session ?? NewSession(connStr, password, passcode, sessionOrCreateTokens.SessionCreationToken());
             if (sessionOrCreateTokens.Session != null)
             {
                 _sessionPoolEventHandler.OnSessionProvided(this);
             }
-            ScheduleNewIdleSessions(connStr, password, sessionOrCreateTokens.BackgroundSessionCreationTokens());
+            ScheduleNewIdleSessions(connStr, password, RegisterSessionCreationsToEnsureMinPoolSize());
             WarnAboutOverridenConfig();
-            return sessionOrCreateTokens.Session ?? NewSession(connStr, password, sessionOrCreateTokens.SessionCreationToken());
+            return session ?? sessionOrCreateTokens.Session ?? NewSession(connStr, password, passcode, sessionOrCreateTokens.SessionCreationToken());
         }
 
-        internal async Task<SFSession> GetSessionAsync(string connStr, SecureString password, CancellationToken cancellationToken)
+        private void ValidatePoolingIfPasscodeProvided(SFSessionProperties sessionProperties)
+        {
+            if (!GetPooling() || _poolConfig.MinPoolSize == 0) return;
+            var isUsingPasscode = (sessionProperties.IsNonEmptyValueProvided(SFSessionProperty.PASSCODE) ||
+                                   (sessionProperties.TryGetValue(SFSessionProperty.PASSCODEINPASSWORD, out var passcodeInPasswordValue) &&
+                                    bool.TryParse(passcodeInPasswordValue, out var isPasscodeinPassword) && isPasscodeinPassword));
+            var isMfaAuthenticator = sessionProperties.TryGetValue(SFSessionProperty.AUTHENTICATOR, out var authenticator) &&
+                                     authenticator == MFACacheAuthenticator.AUTH_NAME;
+            if(isUsingPasscode && !isMfaAuthenticator)
+            {
+                const string ErrorMessage = "Could not use connection pool because passcode was provided using a different authenticator than username_password_mfa";
```

Comment:
actually the problem is not using the pool but using minPoolSize feature

Reply:
Do you have any recommendation for a better  error message?

Reply:
maybe `Passcode with MinPoolSize feature of connection pool allowed only for username_password_mfa authentication`?

Reply:
Done